crates/stateless/src/validation.rs

```diff
@@ -0,0 +1,201 @@
+use std::{collections::HashMap, sync::Arc};
+
+use crate::witness_db::WitnessDatabase;
+use alloy_consensus::{Block, BlockHeader, Header};
+use alloy_primitives::{keccak256, map::B256Map, B256};
+use alloy_rpc_types_debug::ExecutionWitness;
+use reth_chainspec::ChainSpec;
+use reth_ethereum_consensus::validate_block_post_execution;
+use reth_evm::execute::{BlockExecutorProvider, Executor};
+use reth_evm_ethereum::execute::EthExecutorProvider;
+use reth_primitives::{RecoveredBlock, TransactionSigned};
+use reth_revm::state::Bytecode;
+use reth_trie::{iter::IntoParallelRefIterator, HashedPostState, KeccakKeyHasher};
+use reth_trie_sparse::{blinded::DefaultBlindedProviderFactory, SparseStateTrie};
+
+/// Performs stateless validation of a block using the provided witness data.
+///
+/// This function attempts to fully validate a given `current_block` statelessly, ie without access
+/// to a persistent database.
+/// It relies entirely on the `witness` data and `ancestor_headers`
+/// provided alongside the block.
+///
+/// The witness data is validated in the following way:
+///
+/// 1. **Ancestor Header Verification:** Checks if the `ancestor_headers` are present, form a
+///    contiguous chain back from `current_block`'s parent, and do not exceed the `BLOCKHASH` opcode
+///    limit using [`compute_ancestor_hashes`]. We must have at least one ancestor header, even if
+///    the `BLOCKHASH` opcode is not used because we need the state root of the previous block to
+///    verify the pre state reads.
+///
+/// 2. **Pre-State Verification:** Retrieves the expected `pre_state_root` from the parent header
+///    from `ancestor_headers`. Verifies the provided [`ExecutionWitness`] against this root using
+///    [`verify_execution_witness`].
+///
+/// 3. **Chain Verification:** The code currently does not verify the [`ChainSpec`] and expects a
+///    higher level function to assert that this is correct by, for example, asserting that it is
+///    equal to the Ethereum Mainnet Chainspec or asserting against the genesis hash that this
+///    ChainSpec defines.
+///
+/// High Level Overview of functionality:
+///
+/// - Verify all state accesses against a trusted pre-state root
+/// - Put all state accesses into an in-memory database
+/// - Use the in-memory database to execute the block
+/// - Validate the output of block execution (e.g. receipts, logs, requests)
+/// - Compute the post-state root using the state-diff from block execution
+/// - Check that the post-state root is the state root in the block.
+///
+/// If all steps succeed the function returns `Some` containing the hash of the validated
+/// `current_block`.
+// TODO(Note): The code currently unwraps in a lot of places. This will be cleaned up.
+pub fn stateless_validation(
+    current_block: RecoveredBlock<Block<TransactionSigned>>,
+    witness: ExecutionWitness,
+    ancestor_headers: Vec<Header>,
+    chain_spec: Arc<ChainSpec>,
```

Comment:
Passing note that with stateless(verkle) -- the execution witness is in the block and the block now contains a pre-state root.

The pre-state root being in the block means that we can remove the restriction on us needing at least one ancestor header. This is only needed to get the pre-state root from the parent block. Moreover, we would not need ancestor_headers for verkle due to 7709/2935